### PR TITLE
feat/GAUSSIAN-FRAME-31

### LINKED ISSUE #31: Implement gaussian input function

### PR DESCRIPTION
**Pull request by universvm**
Closes #31

### CODE REVIEW

**Review by ChrisWellsWood — CHANGES_REQUESTED** [review]

**Review by ChrisWellsWood — COMMENTED** [review]
Looks excellent, great job Leo. Just a few minor points or queries. I'm not going to approve this yet as you and @MichaelJamesStam are due to have your meeting to discuss this later in the week. Well done again, this was clearly a lot of work.

File: src/aposteriori/data_prep/create_frame_data_set.py [hunk 1]
```diff
@@ -22,10 +23,12 @@
 
 from ampal.amino_acids import standard_amino_acids
 from aposteriori.config import (
-    UNCOMMON_RESIDUE_DICT,
+    ATOMIC_CENTER,
+    ATOM_WANDERWAAL_RADII,
```
[the remark this answers is not in the document]
Comment (universvm) [reply src/aposteriori/data_prep/create_frame_data_set.py:27]: oops! Thanks for spotting it.

File: src/aposteriori/config.py [hunk 2]
```diff
@@ -23,4 +29,4 @@
...
 UNCOMMON_RES_CONVERSION = True
-
+ATOMIC_CENTER = (1, 1, 1)
```
[the remark this answers is not in the document]
Comment (universvm) [reply src/aposteriori/config.py:32]: It depends on whether the frame is 3x3x3 or not. My thought was that by adding it as a config variable we could  expand to something slightly larger than that.

File: src/aposteriori/config.py [hunk 3]
```diff
@@ -15,6 +15,12 @@
     3: "orange",  # +1
     4: "green"  # +2
 }
+ATOM_WANDERWAAL_RADII = {
+    # Atomic number : Radius
+    0: 0.7,  # Carbon
+    1: 0.65,  # Nitrogen
+    2: 0.6,  # Oxygen
+}
```
[the remark this answers is not in the document]
Comment (universvm) [reply src/aposteriori/config.py:23]: We could change it to a list actually. I just wanted to be explicit about the key.

File: src/aposteriori/data_prep/create_frame_data_set.py [hunk 4]
```diff
@@ -282,12 +353,190 @@ def encode_residue(residue: str) -> np.ndarray:
...
+
+    Returns
+    -------
+    norm_gaussian_frame: np.ndarray
+        3x3x3 Frame encoding for a gaussian atom
+    """
+    # Unpack x, y, z:
+    x, y, z = modifiers_triple
+    # Obtain x, y, z ranges for the gaussian
+    x_range = y_range = z_range = np.linspace(-range_val, range_val, resolution)
```
[the remark this answers is not in the document]
Comment (universvm) [reply src/aposteriori/data_prep/create_frame_data_set.py:385]: Yeah we could go for `gaussian_range`. I thought this step required additional expciciteness hence why I've assigned a value for each axis.

**Comment by universvm** [comment]
As per private conversation with @MichaelJamesStam to more things that can be done are:

- [ ]  Add a test for a frame with gaussians 

- [ ]  Do an unoptimized version of `convert_atom_to_gaussian_density` and check that the they both produce the same gaussian frame

File: src/aposteriori/config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChrisWellsWood) [reply src/aposteriori/config.py:32]: As we mentioned earlier, I'd remove this and hard code it into the function.

File: src/aposteriori/data_prep/create_frame_data_set.py [hunk 5]
```diff
@@ -130,6 +134,71 @@ def encode_atom(self, atom_label: str) -> np.ndarray:
...
+                atom_label = "C"
+                atom_idx = self.label_to_encoding[atom_label]
+        elif self.encoder_length == 4:
+            # In this scenario, Ca is a carbon atom but Cb has a separate channel
+            if atom_label.upper() == "CA":
+                atom_label = "C"
+                atom_idx = self.label_to_encoding[atom_label]
+        else:
+            raise ValueError(
+                f"{atom_label} not found in {self.atomic_labels} encoding. Returning empty array."
```
Comment (ChrisWellsWood) [inline src/aposteriori/data_prep/create_frame_data_set.py:178]: I think the "Returning" part is left over from when it was an error. This won't return anything, as execution is halted by the error unless caught in a `try` statement.

File: src/aposteriori/data_prep/create_frame_data_set.py [hunk 6]
```diff
@@ -282,12 +351,213 @@ def encode_residue(residue: str) -> np.ndarray:
     return residue_encoding
 
 
+def convert_atom_to_gaussian_density(
+    modifiers_triple: t.Tuple[float, float, float],
+    wanderwaal_radius: float,
+    range_val: int = 2,
+    resolution: int = 999,
+    optimized: bool = True,
```
Comment (ChrisWellsWood) [inline src/aposteriori/data_prep/create_frame_data_set.py:359]: You don't have any documentation for this flag here.

File: src/aposteriori/data_prep/create_frame_data_set.py [hunk 7]
```diff
@@ -937,6 +1261,9 @@ def make_frame_dataset(
         If True, the user will be prompted to start creating the dataset.
     encode_cb: bool
         Whether to encode the Cb at an average position in the frame.
+    voxels_as_gaussian: bool
+        Whether to encode voxels as gaussians.
```
Comment (ChrisWellsWood) [inline src/aposteriori/data_prep/create_frame_data_set.py:1265]: As this is the top level function, it is the function that people are most likely to interact with, so it'd be good to have a fully description here, maybe with the citation or explaining the problem it attempts to solve.

**Review by ChrisWellsWood — DISMISSED** [review]
This looks great, happy to approve. One thing that popped into my head at the end though, do we have tests for the _original_ (non-gaussian) format? We need protect against a regression.

**Review by ChrisWellsWood — APPROVED** [review]

**Pull request merged**